rfcs/2020-06-18-2625-architecture-revisit.md

```diff
@@ -0,0 +1,504 @@
+far. They also don't currently appear to be much of a performance bottleneck, so
+the concerns here are somewhat independent of (or at least not as urgent as)
+those above.
+
+The most obvious problem with our current sinks is that we directly reuse the
+`Sink` trait from the `0.1` version of the `futures` crate. This will obviously
+need to change before we can complete our transition to the `0.3` version of
+`futures`.
+
+One option would be to simply transition to the `0.3` version of the trait, but
+that is undesirable for a number of reasons. First, the new version is
+significantly more complex. Second, there is a general sentiment in the
+community that the trait is not worth the complexity and is unlikely to be
+stabilized in the way that others have.
+
+Instead, we should migrate towards our own type that both insulates us from this
+ecosystem uncertainty and more closely aligns with our specific needs.
+
+```rust
+trait SinkConfig {
+    async fn build(&self) -> Result<Sink>;
+}
+
+enum Sink {
+    Streaming(StreamingSink),
+    Service(ServiceSink),
+    OldAndBad(futures01::Sink),
+}
+
+impl Sink {
+    async fn run(self, input: Stream<Event>) -> Result<()> {
+        ...
+    }
+}
+```
+
+Again, this is just pseudocode, but it illustrates a couple of important points:
+
+1. There are a number of types of sinks that can differ in implementation details
+2. We provide a simple, unified, stream-based interface to drive all of them
+
+The most important benefit here is that we don't need to rewrite everything at
+once to move to a new style. You can simply wrap `Sink::OldAndBad` around
+existing implementation and upgrade them later one at a time.
+
+Another important point is that we've decoupled the topology-level API from the
+implementation-level API. This means we can start with something simple like the
+`run` method above, and later move towards something more sophisticated without
+necessarily touching every individual component.
+
+Once this seam is established, the next most important abstraction is
+`ServiceSink`. We have something similar right now in `BatchSink`, but we've
+begun to see limitations of that design. A full proposal is outside of the scope
+of this RFC, but future work should explore the relationship between batching,
+encoding, and request building. Their separation in the current design appears
+to be the cause of some limitations and there's likely an improvement to be made
+by evaluating and adjusting to their mutual dependencies.
+
+### Topology design
+
+With the above component changes in place, we'll gain some flexibility in how we
+build topologies. The goal will be, where possible, to consolidate stateless
+transform-like processing into cloneable `Push` implementations that are passed
+to sources in place of the current raw channels senders. This will effectively
+"inline" this logic into the source task, allowing it to execute with the
+natural concurrency of a given source (e.g. per connection).
```

Comment:
This might be a dumb question, but did you think about inlining sinks as well? I realize this raises a number of questions that are not worth the complexity right now but could this further improve performance for compatible pipelines?

For example:

```
            socket source
|-----------------------------------|
 [conn 1] - [transform 1] - [sink 1]
 [conn 2] - [transform 1] - [sink 1]
 [conn 3] - [transform 1] - [sink 1]
|-----------------------------------|
```

Instead of:

```
       socket source
|------------------------|
 [conn 1] - [transform 1] \
 [conn 2] - [transform 1] - [sink 1]
 [conn 3] - [transform 1] /
|------------------------|
```

Does that make sense?

Reply:
I'm also realizing that tasks are ephemeral, which complicates how these sinks would clean up in the process. I'm mostly curious from a theoretical standpoint.

Reply:
My rough idea here is that sinks eventually be decomposed into smaller component parts, some of which would be compatible with inlining. For example, encoding an event to JSON is perfectly fine to inline, but batching is fundamentally something that requires coordination. Partitioned batching is an interesting case too, where we could potentially flow that information back up the graph and use it to parallelize the whole pipeline.